src/storage/prisma.ts

```diff
@@ -289,32 +274,53 @@ export class PrismaStorage implements IDriveStorage {
             header: DocumentHeader;
             newState?: State<any, any> | undefined;
         }>
-    ) {
+    ): Promise<void> {
         let result: {
             operations: Operation[];
             header: DocumentHeader;
             newState?: State<any, any> | undefined;
         } | null = null;
 
-        await this.db.$transaction(
-            async tx => {
-                const document = await this.getDocument(drive, id, tx);
-                if (!document) {
-                    throw new Error(`Document with id ${id} not found`);
+        try {
+            await this.db.$transaction(
+                async tx => {
+                    const document = await this.getDocument(drive, id, tx);
+                    if (!document) {
+                        throw new Error(`Document with id ${id} not found`);
+                    }
+                    result = await callback(document);
+
+                    const { operations, header, newState } = result;
+                    return this._addDocumentOperations(
+                        tx,
+                        drive,
+                        id,
+                        operations,
+                        header
+                    );
+                },
+                {
+                    isolationLevel: 'ReadCommitted',
+                    maxWait: 10000,
+                    timeout: 20000
                 }
-                result = await callback(document);
-
-                const { operations, header, newState } = result;
-                return this._addDocumentOperations(
-                    tx,
+            );
+        } catch (e) {
+            // if optimistic update fails, retry with new state
+            if (
+                e instanceof PrismaClientKnownRequestError &&
+                e.code === 'P2025'
+            ) {
+                // TODO add exponential backoff?
+                return this.addDocumentOperationsWithTransaction(
```

Comment:
Retries the entire process again, reads the new operations and syncUnit version, applies the shuffle if needed and writes again.

We probably should add an exponential backoff to avoid overloading the db when there are a lot of requests to the same sync unit. But this should be done at the queue level.